### PR TITLE
ICU-22736 Improve Persian Calendar

### PR DESCRIPTION
**Pull request by FrankYFTang**
##### Checklist

- [X] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22736
- [X] Required: The PR title must be prefixed with a JIRA Issue number. 
- [X] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [X] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [X] Issue accepted (done by Technical Committee after discussion)
- [X] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

ALLOW_MANY_COMMITS=true

### CODE REVIEW

File: icu4c/source/test/intltest/incaltst.cpp [hunk 1]
```diff
@@ -896,6 +900,429 @@ void IntlCalendarTest::TestPersian() {
...
+      {739331, 1404, 1, 1},
+      {744313, 1417, 8, 19},
+      {763436, 1469, 12, 30},
+      {763437, 1470, 1, 1},
+      {764652, 1473, 4, 28},
+      {775123, 1501, 12, 29},
+      {775488, 1502, 12, 29},
+      {775487, 1502, 12, 28},
+      {775488, 1502, 12, 29},
+      {775489, 1503, 1, 1},  // real
```
Comment (roozbehp) [inline icu4c/source/test/intltest/incaltst.cpp:930]: What does this comment, "real", mean here?

File: icu4j/main/core/src/test/java/com/ibm/icu/dev/test/calendar/PersianTest.java [hunk 2]
```diff
@@ -143,4 +144,407 @@ public StubCalendar() {
...
+        739331, 1404, 1, 1,
+        744313, 1417, 8, 19,
+        763436, 1469, 12, 30,
+        763437, 1470, 1, 1,
+        764652, 1473, 4, 28,
+        775123, 1501, 12, 29,
+        775488, 1502, 12, 29,
+        775487, 1502, 12, 28,
+        775488, 1502, 12, 29,
+        775489, 1503, 1, 1,  // real
```
Comment (roozbehp) [inline icu4j/main/core/src/test/java/com/ibm/icu/dev/test/calendar/PersianTest.java:170]: Same here.

**Review by roozbehp — APPROVED** [review]

**Review by roozbehp — APPROVED** [review]

**Comment by FrankYFTang** [comment]
@sffc I need an approved reviewer to approve

File: icu4c/source/i18n/persncal.cpp [hunk 3]
```diff
@@ -62,6 +65,45 @@ static const int32_t kPersianCalendarLimits[UCAL_FIELD_COUNT][4] = {
...
+       2063, 2096, 2129, 2158, 2162, 2191, 2195, 2224, 2228, 2257, 2261, 2290, 2294, 2323, 2327, 2356,
+       2360, 2389, 2393, 2422, 2426, 2455, 2459, 2488, 2492, 2521, 2525, 2554, 2558, 2587, 2591, 2620,
+       2624, 2653, 2657, 2686, 2690, 2719, 2723, 2748, 2752, 2756, 2781, 2785, 2789, 2818, 2822, 2847,
+       2851, 2855, 2880, 2884, 2888, 2913, 2917, 2921, 2946, 2950, 2954, 2979, 2983, 2987,
+    };
+    gMinCorrection = nonLeapYears[0];
+    icu::UnicodeSet prefab;
+    for (int32_t i = 0; i < UPRV_LENGTHOF(nonLeapYears); i++) {
+        prefab.add(nonLeapYears[i]);
+    }
```
Comment (markusicu) [inline icu4c/source/i18n/persncal.cpp:98]: optional
```suggestion
    for (auto year : nonLeapYears) {
        prefab.add(year);
    }
```

File: icu4j/main/core/src/main/java/com/ibm/icu/util/PersianCalendar.java [hunk 4]
```diff
@@ -93,6 +95,22 @@ public class PersianCalendar extends Calendar {
...
+       2624, 2653, 2657, 2686, 2690, 2719, 2723, 2748, 2752, 2756, 2781, 2785, 2789, 2818, 2822, 2847,
+       2851, 2855, 2880, 2884, 2888, 2913, 2917, 2921, 2946, 2950, 2954, 2979, 2983, 2987,
+    };
+
+    private static Set<Integer> LEAP_CORRECTION = null;
+    {
+        Set<Integer> prefab = new HashSet<Integer>(NON_LEAP_YEARS.length);
+        for (int nonLeap : NON_LEAP_YEARS) {
+            prefab.add(nonLeap);
+        }
```
Comment (markusicu) [inline icu4j/main/core/src/main/java/com/ibm/icu/util/PersianCalendar.java:111]: I think this means that every time you do a lookup, you end up auto-boxing the year to an Integer object. And I think these are large enough not to be cached.

How about using a [java.util.BitSet](https://docs.oracle.com/javase/8/docs/api/java/util/BitSet.html)?
You could minimize it by also keeping the min value, and storing and looking up (year-minYear).

Something like
```
private static final class NonLeapYears {
    int minYear = NON_LEAP_YEARS[0];
    int maxYear = NON_LEAP_YEARS[NON_LEAP_YEARS.length - 1];
    BitSet offsetYears;

    NonLeapYears() {
        offsetYears = new BitSet(maxYear - minYear + 1);
        for (int nonLeap : NON_LEAP_YEARS) {
            offsetYears.set(nonLeap - minYear);
        }
    }

    boolean contains(int year) {
        return minYear <= year && year <= maxYear && offsetYears.get(year - minYear);
    }
}
```
and then here you initialize LEAP_CORRECTION = new NonLeapYears();
and below you just call LEAP_CORRECTION.contains(year)
without an additional year range check

WDYT?

File: icu4c/source/i18n/persncal.cpp [hunk 5]
```diff
@@ -62,6 +65,45 @@ static const int32_t kPersianCalendarLimits[UCAL_FIELD_COUNT][4] = {
...
+       2360, 2389, 2393, 2422, 2426, 2455, 2459, 2488, 2492, 2521, 2525, 2554, 2558, 2587, 2591, 2620,
+       2624, 2653, 2657, 2686, 2690, 2719, 2723, 2748, 2752, 2756, 2781, 2785, 2789, 2818, 2822, 2847,
+       2851, 2855, 2880, 2884, 2888, 2913, 2917, 2921, 2946, 2950, 2954, 2979, 2983, 2987,
+    };
+    gMinCorrection = nonLeapYears[0];
+    icu::UnicodeSet prefab;
+    for (auto year : nonLeapYears) {
+        prefab.add(year);
+    }
+    gLeapCorrection = prefab.cloneAsThawed();
```
Comment (markusicu) [inline icu4c/source/i18n/persncal.cpp:99]:
```suggestion
    gLeapCorrection = prefab.cloneAsThawed().freeze();
```

should make contains() a lot faster (and makes the set immutable)

https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/classicu_1_1UnicodeSet.html#ad8d781d46d63eb57eb24e372feef12ef

**Review by markusicu — APPROVED** [review]

**Pull request merged**